src/javascript/cons_list.ts

```diff
@@ -0,0 +1,82 @@
+// A type-safe immutable homogenous "persistent" cons list.
+// A cons list is either empty or a cons pair of T and a cons list.
+export type ConsList<T> = Cons<T>|Empty;
+
+// Creates a non-empty cons list.
+export function cons<T>(head: T, tail: ConsList<T> = empty()): Cons<T> {
+  return new Cons<T>(head, tail);
+}
+
+// Returns the empty cons list.
+export function empty(): Empty {
+  return Empty.instance;
+}
+
+// Methods implemented by both Empty and Cons<T>.
+export interface ConsListInterface<T> extends Iterable<T> {
+  // Check whether the list is an Empty.
+  //
+  // Using this method is preferable to the expensive and less readable
+  // `instanceof Empty` check.
+  isEmpty(): this is Empty;
+
+  // Equivalent to [...list] but slightly faster.
+  toArray(): T[];
```

Comment:
I'm not seeing any usages of ConsList at the moment, but it's weird that you don't have any other ways to use it than pattern-match and isEmpty and toArray functions in the interface.

Pattern matching should be fine, but IMO it would be useful to have a reduceRight in the interface as well.

Reply:
See `waxeye.ts` diff for usages. Nothing else is needed here.